File: Days.34/C++/1.Construct-BST-from-Level-Order-Traversal.cpp
// Construct BST from Level Order Traversal 

#include<iostream>
#include<vector>
#include<queue>
using namespace std;

class Node
{
    public:
    int val;
    Node *left;
    Node *right;
    Node(int val)
    {
        this->val=val;
        this->left=this->right=nullptr;
    }
};

class Pair
{
    public:
    Node *root;
    int mini;
    int maxi;
};

Node* BuildBST(int n,vector<int> &level)
{
    Node *curr;
    int maxi;
    int mini;
    Node *root=new Node(level[0]);
    Pair p,temp;
    p.root=root;
    p.mini=INT_MIN;
    p.maxi=INT_MAX;
    queue<Pair> q;
    q.push(p);
    int i=1;
    while (i<n and !q.empty())
    {
        p=q.front();
        q.pop();
        curr=p.root;
        maxi=p.maxi;
        mini=p.mini;
        if (i<n and level[i]>mini and level[i]<maxi and level[i]<curr->val)
        {
            curr->left=new Node(level[i]);
            temp.root=curr->left;
            temp.mini=mini;
            temp.maxi=curr->val;
            i++;
            q.push(temp);
        }
        if(i<n and level[i]>mini and level[i]<maxi and level[i]>curr->val)
        {
            curr->right=new Node(level[i]);
            temp.root=curr->right;
            temp.mini=curr->val;
            temp.maxi=maxi;
            i++;
            q.push(temp);
        }
    }
    return root;
}

void Inorder(Node *root)
{
    if(root==nullptr)
    {
        return;
    }
    Inorder(root->left);
    cout<<root->val<< " ";
    Inorder(root->right);
}

int main()
{
    int n;
    cin>>n;
    vector<int> level(n);
    for(int i=0; i<n; i++)
    {
        cin>>level[i];
    }
    Node *root=BuildBST(n,level);
    Inorder(root);
    cout<<"\n";
    return 0;
}

File: Days.34/Python/1.Construct-BST-from-Level-Order-Traversal.py
# Construct BST from its Level Order Traversal

from collections import deque

class Node:
    def __init__(self,val):
        self.val=val
        self.left=None
        self.right=None

def BuildBST(n,level):
    dq=deque()
    root=Node(level[0])
    dq.append([root,-2**63,2**63])
    i=1
    while i<n and dq:
        curr,mini,maxi=dq.popleft()
        if i<n and level[i]>mini and level[i]<maxi and level[i]<curr.val:
            curr.left=Node(level[i])
            dq.append([curr.left,mini,curr.val])
            i+=1
        if i<n and level[i]>mini and level[i]<maxi and level[i]>curr.val:
            curr.right=Node(level[i])
            dq.append([curr.right,curr.val,maxi])
            i+=1
    return root

def Inorder(root):
    if root==None:
        return
    Inorder(root.left)
    print(root.val,end= " ")
    Inorder(root.right)

def main():
    n=int(input())
    level=list(map(int,input().split()))
    root=BuildBST(n,level)
    Inorder(root)
    print()



main()

